grep-0018-modernize-c++.md

```diff
@@ -0,0 +1,417 @@
+manual memory management, improving readability, and reducing chances
+for mistakes like copy-paste ` * sizeof(float)` where ` *
+sizeof(gr_complex)` was intended.
+
+Recommendation:
+* Replace all feasible uses of `volk_malloc` et al with `volk::vector`
+
+Risks: master branch already depends on `volk::vector`, so none.
+
+### `xxx_cast<>`
+
+`static_cast<>` and friends are safer than C style casts, but more
+verbose.
+
+A typical example is:
+
+```c++
+[…]
+  gr_vector_const_void_star& input_items,
+[…]
+  float* in = (float*)input_items[0];
+```
+
+which should be changed to use a `static_cast<>` in order to not cast
+away the constness.
+
+Recommendation:
+* Change them where they make sense, but don't have as a goal to
+  eliminate C style casts where they seem right
+
+Risks: none
+
+### `static_assert`
+
+There are many uses of `assert()` (runtime check) that should be
+`static_assert` (compile time check).
+
+Example:
+
+```c++
+assert(sizeof(fftwf_complex) == sizeof(gr_complex))
+```
+
+Risks: none
+
+### Default member initialization
+
+Pre-C++11 there were only two ways to initialize values, both in the
+constructor:
+
+```c++
+class obj {
+  int d_a, d_b, d_c;
+  std::string d_s;
+public:
+  obj(int a)
+    : d_a(a),
+      d_b(0),
+      d_s()   // Completely unnecessary
+  {
+      d_c = 3;
+  }
+  obj(int a, int b)
+    : d_a(a), d_b(b)  // Repeated init of d_a.
+  {
+      d_c = 3; // Repeated init of d_c;
+  }
+};
+```
+
+This causes code duplication.
+
+In C++11 default member values can be set at declaration time
+(normally in header files):
+
+```c++
+class obj {
+  int d_a = 0;   // Or int d_a{0};
+  int d_b = 0;
+  int d_c = 3;
+public:
+  obj(int a)
+    : d_a(a)
+  {
+  }
+  obj(int a, int b):
+    : d_a(a), d_b(b)    // Duplication of d_a(a).
+  {
+  }
+};
+```
+
+To completely avoid duplication one constructor can call another:
+
+```c++
+class obj {
+  int d_a = 0;   // Or int d_a{0};
+  int d_b = 0;
+  int d_c = 3;
+public:
+  obj(int a)
+    : d_a(a)
+  {
+  }
+  obj(int a, int b)
+    : obj(a) // Call the other constructor.
+  {
+    d_b = b; // Can't be put in construct initializer when
+             // construction delegation is used.
+  }
+};
+```
+
+Recommendation:
+
+* Don't use constructor delegation with a non-empty body. It's
+  semantically unclear what it means when object is constructed, but a
+  constructor is still running.
+* Avoid initializing inside the constructor body, especially when it
+  prevents a member being const.
+* Don't needlessly list member variables for their default constructor
+  (`d_s`, above)
+* Use declaration initialization for default values. It's the least
+  repetition in the face of multiple constructors.
+
+Risk: This change does not affect API, so none.
+
+### `enum class`
+
+Convert all `enum` to type-safe `enum class`.
+
+Recommendation:
+* Internal enums can be changed freely at any time
+* enums part of API can be changed between minor releases.
+
+Risks: none
```

Comment:
This carries a huge risk, since the API to access all enums would change.

Reply:
Risk even if done between minor versions?

Reply:
You mean between major versions? I would say yes, because it invalidates OOT all over the place. Sure, OOT maintainers can update, but this kind of change is annoying to automate.

I would suggest changing this section, or even ditching it. New enums should be enum classes, but I see no need to modify existing enums.

Reply:
I meant the "API" digit. B in A.B.C.D. 

[This](https://lists.gnu.org/archive/html/discuss-gnuradio/2018-02/msg00133.html) doesn't say, but I guess it's Paradigm.Major.Minor.Patch. So yes, Major.

OK, dropping changing these where it's part of the API. I would argue for changing it in API code too though, because of the extra compiler support (warnings) it adds, and forced type safety.

Reply:
Please re-check the abstract for more motivation, as the change I just made seems incompatible with the Abstract as currently written.